api/src/main/java/org/openmrs/module/fhir/api/impl/PersonServiceImpl.java

```diff
@@ -92,4 +96,65 @@ public Person createFHIRPerson(Person person) {
 		return FHIRPersonUtil.generatePerson(omrsPerson);
 	}
 	
+	@Override
+	public Person updateFHIRPerson(Person thePerson, String theId) {
+		org.openmrs.api.PersonService personService = Context.getPersonService();
+		org.openmrs.Person retrievedPerson = personService.getPersonByUuid(theId);
+		if (retrievedPerson != null) { // update person
+			org.openmrs.Person omrsPerson = FHIRPersonUtil.generateOpenMRSPerson(thePerson);
+			retrievedPerson = updatePersonAttributes(omrsPerson, retrievedPerson);
+			Context.getPersonService().savePerson(retrievedPerson);
+			return FHIRPersonUtil.generatePerson(retrievedPerson);
+		} else { // no person is associated with the given uuid. so create a new person with the given uuid
+			if (thePerson.getId() == null) { // since we need to PUT the Person to a specific URI, we need to set the uuid here, if it is not already set.
+				IdDt uuid = new IdDt();
+				uuid.setValue(theId);
+				thePerson.setId(uuid);
+			}
+			return createFHIRPerson(thePerson);
+		}
+	}
+	
+	public org.openmrs.Person updatePersonAttributes(org.openmrs.Person omrsPerson, org.openmrs.Person retrievedPerson) {
+		Set<PersonName> all = retrievedPerson.getNames();
+		boolean needToSetPrefferedName = false; // indicate wheter any preffered names are in the request body. 
+		for (PersonName name : omrsPerson.getNames()) {
+			if (name.getPreferred()) { // detecting any preffered names are in the request body
+				needToSetPrefferedName = true;
+			}
+		}
+		if (needToSetPrefferedName) { // unset the existing preffered name, 
+			for (PersonName name : all) {
+				name.setPreferred(false);
```

Comment:
@harsha89 , I used the approach used in webservicerest module https://github.com/openmrs/openmrs-module-webservices.rest/blob/93b2779e051158c1e43e0a601303b5f2aacedbfe/omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java#L160

we need to do 
if(name.isPreferred) {
then name.setPreferred(false)
}
only if we have a prefered name in json body. because we can have only one prefered name  in data model. so in the first for loop I detect whether there any prefered names in the json body. (by needToSetPrefferedName = true;). if it is so, i unset all names in the data model in server(For this not checking, whether a particular name is prefered or not  will not make any harm. because anyways they should be unset).  then I copy all names in the json body to the server. Then the prefered name which was in the json body will be copied to the server. So we have a new prefered name now. 

What do you think?